### PR TITLE
[Merged by Bors] - update system test example to include using events

### LINKED ISSUE #4936: Add example for testing systems that use events.
I couldn't find any good examples of this. Got some help on the Discord and it seems to work, but would be nice to have some clear examples. Could potentially extend [this example](https://github.com/bevyengine/bevy/blob/main/tests/how_to_test_systems.rs).

```rust
fn test() {
    let mut app = App::new();

    app.add_event::<NetworkInput>();
    app.add_event::<NetworkOutput>();
    app.add_system(super::do_things);

    app.world
        .resource_mut::<Events<NetworkInput>>()
        .send(NetworkInput {
            // ...
        });

    app.update();

    let output_events = app.world.resource::<Events<NetworkOutput>>();
    let mut output_reader = output_events.get_reader();
    let output = output_reader.iter(&output_events).next().unwrap();

    // assert on `output`
}
```

### PR DESCRIPTION
**Pull request by its-danny**
# Objective

- Adds an example of testing systems that handle events. I had a hard time figuring out how to do it a couple days ago so figured an official example could be useful.
- Fixes #4936

## Solution

- Adds a `Score` resource and an `EnemyDied` event. An `update_score` system updates the score when a new event comes through. I'm not sure the example is great, as this probably isn't how you'd do it in a real game, but I didn't want to change the existing example too much.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
I'm happy with this! I'd like to merge this before reviving #3839.

File: tests/how_to_test_systems.rs [hunk 1]
```diff
@@ -21,49 +37,76 @@ fn hurt_enemies(mut enemies: Query<&mut Enemy>) {
...
+        .id();
 
     // Run systems
-    update_stage.run(&mut world);
+    app.update();
 
     // Check resulting changes
-    assert!(world.get::<Enemy>(enemy_id).is_some());
-    assert_eq!(world.get::<Enemy>(enemy_id).unwrap().hit_points, 4);
+    assert!(app.world.get::<Enemy>(enemy_id).is_some());
```
Comment (mockersf) [inline tests/how_to_test_systems.rs:76]: why did you remove the assert on the hit points of the entity?

File: tests/how_to_test_systems.rs [hunk 2]
```diff
@@ -21,49 +37,76 @@ fn hurt_enemies(mut enemies: Query<&mut Enemy>) {
...
+        .id();
 
     // Run systems
-    update_stage.run(&mut world);
+    app.update();
 
     // Check resulting changes
-    assert!(world.get::<Enemy>(enemy_id).is_none());
+    assert!(app.world.get::<Enemy>(enemy_id).is_none());
+    assert!(app.world.get::<Enemy>(enemy_id).is_none());
```
Comment (mockersf) [inline tests/how_to_test_systems.rs:109]: this looks like the same assert twice?

File: tests/how_to_test_systems.rs [hunk 3]
```diff
@@ -95,3 +138,40 @@ fn spawn_enemy_using_input_resource() {
...
+    // Run systems
+    app.update();
+
+    // Get `EnemyDied` event reader
+    let enemy_died_events = app.world.resource::<Events<EnemyDied>>();
+    let mut enemy_died_reader = enemy_died_events.get_reader();
+    let enemy_died = enemy_died_reader.iter(enemy_died_events).next().unwrap();
+
+    // Check the event has been sent
+    assert_eq!(enemy_died.0, 3);
```
Comment (mockersf) [inline tests/how_to_test_systems.rs:170]: this is just asserting that the event you sent yourself is sent, it doesn't use anything from the system

**Review by mockersf — CHANGES_REQUESTED** [review]
You updated the tests to use an app instead of a stage, except for the test `spawn_enemy_using_input_resource`. Could you also update it?

You also didn't test that the event is correctly sent by the system `despawn_dead_enemies`, this should probably be asserted in the test `did_despawn_enemy`

File: tests/how_to_test_systems.rs (hunk 1 above)
Comment (its-danny) [reply tests/how_to_test_systems.rs:76]: accident, i assume. recently switched to linux and keep middle mouse pasting random things then deleting stuff on accident lol.

File: tests/how_to_test_systems.rs (hunk 2 above)
Comment (its-danny) [reply tests/how_to_test_systems.rs:109]: same as above.

**Comment by its-danny** [comment]
> You updated the tests to use an app instead of a stage, except for the test `spawn_enemy_using_input_resource`. Could you also update it?

i can, just wasn't sure if I should. you only need to use `App` if you're adding events, otherwise using `World` works. since the example previously used that, I assumed it was the `default` way to test and that using app is the exception. if that's not the case tho, lemme know.

> You also didn't test that the event is correctly sent by the system `despawn_dead_enemies`, this should probably be asserted in the test `did_despawn_enemy`

updated

**Comment by alice-i-cecile** [comment]
In my experience, `App` is needed for all but the most trivial tests. Usually you'll want to be testing that your plugin performs some end-to-end behavior, which requires both a schedule and a world.

**Comment by mockersf** [comment]
> i can, just wasn't sure if I should. you only need to use `App` if you're adding events, otherwise using `World` works. since the example previously used that, I assumed it was the `default` way to test and that using app is the exception. if that's not the case tho, lemme know.

There isn't really a "default way to test" yet... For now I would prefer consistency in the tests over showing how to do it in a simpler case that might not be that useful as @alice-i-cecile noted

File: tests/how_to_test_systems.rs (hunk 1 above)
Comment (mockersf) [reply tests/how_to_test_systems.rs:76]: That's linux for you!

File: tests/how_to_test_systems.rs (hunk 3 above)
Comment (its-danny) [reply tests/how_to_test_systems.rs]: updated. was just trying to show how to read and write events in a test, but moved the read to `did_despawn_enemy`.

**Comment by its-danny** [comment]
works for me, updated.

**Review by mockersf — APPROVED** [review]
